Add more supported simulator debugging options and improve tests

vashworth opened the pull request:
Add more supported simulator debugging options update tests. Also, add tests for other launch arguments.

Fixes https://github.com/flutter/flutter/issues/52676.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

vashworth committed:
add debugging options to simulator, test more debugging flags, add tests for other launch arguements

vashworth commented:
Discussed with @jmagman about improving this code by using a single function that returns the list of launch arguments that can then be used for both the iOS simulator and physical iOS. The upside of this would be less repeated code, and if someone wants to add a new flag in the future, they can add to just one place and won't have to remember to add to both simulators.dart and devices.dart.

vashworth commented:
Also discussed how the simulator version only adds the flags if the `debuggingOptions.debuggingEnabled` is true
https://github.com/flutter/flutter/blob/d40f6d1e681f17e3b7b7290afb4c9b4175b976e7/packages/flutter_tools/lib/src/ios/simulators.dart#L432-L448

Whereas the device version adds them regardless, and only adds 2 specific ones if `debuggingOptions.debuggingEnabled` is true.
https://github.com/flutter/flutter/blob/d40f6d1e681f17e3b7b7290afb4c9b4175b976e7/packages/flutter_tools/lib/src/ios/devices.dart#L359-L363

Since the iOS simulator only runs in debug mode (profile and release mode are not supported), we think it's okay to remove the `debuggingOptions.debuggingEnabled` check for the simulator.

vashworth committed:
refactor iOS launch arguments to use one function for both simulator and physical devices

vashworth committed:
Merge remote-tracking branch 'upstream/master' into debug_options

jmagman commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -915,6 +915,42 @@ class DebuggingOptions {
+    final String dartVmFlags = computeDartVmFlags(this);
+    return <String>[
+      '--enable-dart-profiling',
+      if (environmentType == EnvironmentType.physical) '--disable-service-auth-codes',
+      if (environmentType == EnvironmentType.simulator && disableServiceAuthCodes) '--disable-service-auth-codes',
```
Talked to @jonahwilliams, I'm pretty sure this logic is out of date.  For physical devices this was added in https://github.com/flutter/flutter/pull/49735/files#diff-5a3d215e39f292ac706d495f042ef45fab9f13962022f4e8e4580254e8557d52R322 to always be passed in, because there used to be a fallback where we suggested a port to use to the device, and the tool tried that port first, but didn't have the auth code.  However, that fallback was removed in https://github.com/flutter/flutter/pull/68756/files#diff-5a3d215e39f292ac706d495f042ef45fab9f13962022f4e8e4580254e8557d52L368

That means this can be:
```suggestion
      if (disableServiceAuthCodes) '--disable-service-auth-codes',
```
and if something unexpected breaks, we can revert to the old behavior.

vashworth replied:
Changed in Changed in https://github.com/flutter/flutter/pull/114628/commits/c6b500e5dc835ccba5041ca51119424cba078c0c..

jmagman commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -915,6 +915,42 @@ class DebuggingOptions {
+      '--enable-dart-profiling',
+      if (environmentType == EnvironmentType.physical) '--disable-service-auth-codes',
+      if (environmentType == EnvironmentType.simulator && disableServiceAuthCodes) '--disable-service-auth-codes',
+      if (environmentType == EnvironmentType.physical && disablePortPublication) '--disable-observatory-publication',
```
This can also be passed into the simulator, if it's set.  For some reason I thought we didn't do mDNS publication for simulator, but we do:
https://github.com/flutter/engine/blob/a339093f6593e4e155451019af06f333d8f3e325/shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm#L79-L81
https://github.com/flutter/flutter/blob/be83c98edda77ee245764c264752fadf8f2cb8fa/packages/flutter_tools/lib/src/commands/attach.dart#L305-L307
```suggestion
      if (disablePortPublication) '--disable-observatory-publication',
```

vashworth replied:
Changed in https://github.com/flutter/flutter/pull/114628/commits/c6b500e5dc835ccba5041ca51119424cba078c0c.

jmagman commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -915,6 +915,42 @@ class DebuggingOptions {
+      if (cacheSkSL) '--cache-sksl',
+      if (purgePersistentCache) '--purge-persistent-cache',
+      if (route != null) '--route=$route',
+      if (environmentType == EnvironmentType.physical && (platformArgs['trace-startup'] as bool? ?? false)) '--trace-startup',
```
I think we can pass`--trace-startup` through for the simulator, too.
```suggestion
      if (platformArgs['trace-startup'] as bool? ?? false) '--trace-startup',
```

vashworth replied:
Changed in https://github.com/flutter/flutter/pull/114628/commits/c6b500e5dc835ccba5041ca51119424cba078c0c.

vashworth committed:
treat dart flags differently between physical and simulator

jmagman commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -915,6 +915,42 @@ class DebuggingOptions {
+      if (route != null) '--route=$route',
+      if (environmentType == EnvironmentType.physical && (platformArgs['trace-startup'] as bool? ?? false)) '--trace-startup',
+      if (enableImpeller) '--enable-impeller',
+      if (environmentType == EnvironmentType.simulator) '--observatory-port=${hostVmServicePort ?? 0}',
```
I think it can go back to the behavior in https://github.com/flutter/flutter/pull/49735/files#diff-5a3d215e39f292ac706d495f042ef45fab9f13962022f4e8e4580254e8557d52L334 and pass the port along, if it's been set:
```suggestion
      if (hasObservatoryPort) '--observatory-port=${hostVmServicePort}',
```

vashworth replied:
Should this differ between simulator and physical?

Looking at that change, looks like they used `deviceVmServicePort` https://github.com/flutter/flutter/pull/49735/files#diff-5a3d215e39f292ac706d495f042ef45fab9f13962022f4e8e4580254e8557d52L335

Whereas it's `hostVmServicePort` in original simulator args https://github.com/flutter/flutter/blob/01507bab565056b3dee933e2ff146966e4e8b001/packages/flutter_tools/lib/src/ios/simulators.dart#L465.

Is it possible to have both a `deviceVmServicePort` and `hostVmServicePort` and if so, which should be used if both are provided?

jmagman replied:
Good question, I had to ask @jonahwilliams for clarification here, this behavior has changed over time and not every spot it's used has been kept consistent.

To quote Jonah:
> * Application runs on device, it starts up a VM Service instance. This instance will broadcast on a port + auth code combo.
> * (Optional) If Host/Device are different machines, we forward the device port to a selected host port.
> * (Optional) If dds is enabled, dds connects to the selected host port, then broadcasts on the dds port

For physical devices, the right thing to use is `deviceVmServicePort`.  The confusion is what to do if the "device" and host are really the same, as is the case with desktop apps or the iOS simulator, because there's no need to do  port forwarding if they are the same machines.  So I _think_ `hostVmServicePort` is correct in this case, though it's very confusing and it's likely we have some bugs in here.  

@christopherfujino the host/device port logic could use cleanup, comments, better help text, perhaps variable renaming, and verification we're doing the right thing for all device types.

christopherfujino replied:
> @christopherfujino the host/device port logic could use cleanup, comments, better help text, perhaps variable renaming, and verification we're doing the right thing for all device types.

Agreed, let me make a tracking issue.

christopherfujino replied:
> Agreed, let me make a tracking issue.

https://github.com/flutter/flutter/issues/114845

jmagman commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -915,6 +915,43 @@ class DebuggingOptions {
+      if (environmentType == EnvironmentType.simulator && disableServiceAuthCodes) '--disable-service-auth-codes',
+      if (environmentType == EnvironmentType.physical && disablePortPublication) '--disable-observatory-publication',
+      if (startPaused) '--start-paused',
+      if (environmentType == EnvironmentType.physical && dartVmFlags.isNotEmpty) '--dart-flags="$dartVmFlags"',
```
With quotes: [#33924](https://github.com/flutter/flutter/pull/33924/files#diff-5a3d215e39f292ac706d495f042ef45fab9f13962022f4e8e4580254e8557d52R288)
And without quotes: [#63416](https://github.com/flutter/flutter/pull/63416/files#diff-fbf1a86439c67e75e90a00f32fd2b789323582ced2f4f5cfea2d27a1af18aa30R427)

I know you were going to work on https://github.com/flutter/flutter/issues/114622 next, so I guess leave this to be resolved in the next PR.

jmagman replied:
This could use a comment though to explain why they are different.

jmagman replied:
Nit: bump on putting in a comment, even if it's something like "Wrap dart-flags passed to the device in quotes"?

vashworth commented:
Note about https://github.com/flutter/flutter/pull/114628/commits/cc03fe2d4ad53cc25c0ac6805ac5f2d0baf53a22.

So I've discovered that having the quotes surrounding the `--dart-flags` launch argument matters and differs between simulator and physical iOS devices. 

https://github.com/flutter/flutter/blob/cc03fe2d4ad53cc25c0ac6805ac5f2d0baf53a22/packages/flutter_tools/lib/src/device.dart#L926-L927

I tested this locally by using a single check for the dart flags and tried with and without quotes.
``` if (dartVmFlags.isNotEmpty) '--dart-flags="$dartVmFlags"',```

Command I ran `flutter run -d xxx --dart-flags="--max_profile_depth 255,--foo"`
Expected Result: `Encountered disallowed Dart VM flag: --foo`

Test 3 - with quotes, physical device - PASS
Result: `[VERBOSE-3:switches.cc(441)] Encountered disallowed Dart VM flag: --foo`

Test 4 - with quotes, simulator device - FAIL
Result: `[VERBOSE-3:switches.cc(441)] Encountered disallowed Dart VM flag: "--max_profile_depth 255`

Test 1 - without quotes, physical device - FAIL
Result: `Runner[781:39983] Ignoring flag: true is an invalid value for flag max_profile_depth`

Test 2 - without quotes, simulator device - PASS
Result: `[VERBOSE-3:switches.cc(441)] Encountered disallowed Dart VM flag: --foo`

jmagman commented on packages/flutter_tools/lib/src/ios/devices.dart:
```diff
@@ -348,33 +348,7 @@ class IOSDevice extends Device {
-      if (platformArgs['trace-startup'] as bool? ?? false) '--trace-startup',
-      if (debuggingOptions.enableImpeller) '--enable-impeller',
-    ];
+    final List<String> launchArguments = debuggingOptions.getIOSLaunchArguments(EnvironmentType.physical, route, platformArgs);
```
Add a trailing comma and then this becomes shorter:
```suggestion
    final List<String> launchArguments = debuggingOptions.getIOSLaunchArguments(
      EnvironmentType.physical,
      route,
      platformArgs,
    );
```

vashworth replied:
Changed in https://github.com/flutter/flutter/pull/114628/commits/c6b500e5dc835ccba5041ca51119424cba078c0c.

jmagman commented on packages/flutter_tools/lib/src/ios/simulators.dart:
```diff
@@ -445,27 +445,7 @@ class IOSSimulator extends Device {
-        if (route != null) '--route=$route',
-      ],
-    ];
+    final List<String> args = debuggingOptions.getIOSLaunchArguments(EnvironmentType.simulator, route, platformArgs);
```
```suggestion
    final List<String> launchArguments = debuggingOptions.getIOSLaunchArguments(
      EnvironmentType.simulator,
      route,
      platformArgs,
    );
```

vashworth replied:
Changed in https://github.com/flutter/flutter/pull/114628/commits/c6b500e5dc835ccba5041ca51119424cba078c0c.

vashworth committed:
Simplify some flags between devices.

Change --disable-service-auth-codes to not always be included for physical devices, only if disableServiceAuthCodes is true.
Change --disable-observatory-publication to be used for simulator devices too.
Change --enable-checked-mode & --verify-entry-points to be used if debuggingEnabled is true regardless of device type.
Chnage --trace-startup to be used for simulator devices too.

vashworth commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -915,6 +915,42 @@ class DebuggingOptions {
+      if (environmentType == EnvironmentType.simulator && buildInfo.isDebug) ...<String>[
+        '--enable-checked-mode',
+        '--verify-entry-points',
+      ],
```
Changed in https://github.com/flutter/flutter/pull/114628/commits/c6b500e5dc835ccba5041ca51119424cba078c0c.

jmagman replied:
btw you don't need to comment with the updated hash, I get the GitHub notifications when you update the PR. 🙂

vashworth committed:
Merge remote-tracking branch 'upstream/master' into debug_options

vashworth committed:
fix ios release mode with buildable app startApp test

vashworth committed:
determine observatory-port from deviceVmServicePort and hostVmServicePort

jmagman commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -915,6 +915,39 @@ class DebuggingOptions {
+      if (platformArgs['trace-startup'] as bool? ?? false) '--trace-startup',
+      if (enableImpeller) '--enable-impeller',
+      if (environmentType == EnvironmentType.physical && deviceVmServicePort != null) '--observatory-port=$deviceVmServicePort',
+      if (environmentType == EnvironmentType.simulator && hasObservatoryPort) '--observatory-port=$hostVmServicePort',
```
Nit: I know this was my suggestion, but I suspect `hasObservatoryPort` is left over from one of those transitions that needs to be cleaned up (there's a deprecated `flutter run --observatory-port` flag), can you make this null check explicitly for `hostVmServicePort`? This could also use a comment
```suggestion
      // The simulator "device" is actually on the host machine so no ports will be forwarded. Use the suggested host port.
      if (environmentType == EnvironmentType.simulator && hostVmServicePort != null) 
        '--observatory-port=$hostVmServicePort',
```

vashworth replied:
Should I go ahead and remove `hasObservatoryPort`? It's only used on other place.
https://github.com/flutter/flutter/blob/1ca2e0b5b28eb34cb010e8d115f6a4c740de10a0/packages/flutter_tools/lib/src/tester/flutter_tester.dart#L175-L176

jmagman replied:
I think it's wise to remove it, it's not adding that much and is just confusing.

jmagman reviewed (commented):
This LGTM with minor nits

vashworth committed:
add comments and remove hasObservatoryPort

jmagman reviewed (approved):
Thanks for tracking down all the weird corner cases.
